rebuild/src/functions.cc

```diff
@@ -0,0 +1,591 @@
+#include "functions.h"
+
+namespace {
+Napi::Value AggregatePublicKeys(const Napi::CallbackInfo &info) {
+    BLST_TS_FUNCTION_PREAMBLE
+    if (!info[0].IsArray()) {
+        Napi::TypeError::New(env, "publicKeys must be of type PublicKeyArg[]")
+            .ThrowAsJavaScriptException();
+        return scope.Escape(env.Undefined());
+    }
+    Napi::Array arr = info[0].As<Napi::Array>();
+    uint32_t length = arr.Length();
+    if (length == 0) {
+        Napi::TypeError::New(env, "PublicKeyArg[] must have length > 0")
+            .ThrowAsJavaScriptException();
+        return scope.Escape(env.Undefined());
+    }
+
+    BLST_TS_CREATE_JHEAP_OBJECT(wrapped, public_key, PublicKey, result)
+    result->_has_jacobian = true;
+    result->_jacobian.reset(new blst::P1);
```

Comment:
what does this exactly do? initialize with a new P1? and when the result will be garbage collected so will be this?

Reply:
Yes exactly!!  The smart pointers are initially set to `nullptr` on class creation to save space of having unused points and when jacobian or affine are instantiated the point is created once in the heap.  I used smart pointers everywhere to manage the points.  Allows them to be available off JS thread in libuv context without copying.  Only the pointer is copied and ownership stays with the JS object.  They do the cleanup (what makes a pointer "smart") once the object goes out of scope and is garbage collected.

Reply:
ok, so pointers are created in javascript heap, but the actual points (for e.g. P1) is created in c++ heap, and when pointers are garbage collected so does the object?

Reply:
Yep.  Exactly.  Automagically... well i suppose smart-pointedly but its still pretty magic